src/OpenTelemetry/Metrics/MetricAggregators/SumMetricAggregator.cs

```diff
@@ -58,90 +68,224 @@ internal SumMetricAggregator(string name, string description, string unit, Meter
 
         public IEnumerable<IExemplar> Exemplars { get; private set; } = new List<IExemplar>();
 
+        public long Count => this.countLong + this.countDouble;
+
         public IDataValue Sum
         {
             get
             {
-                if (this.valueType == typeof(long))
+                if (this.countDouble == 0)
                 {
                     return new DataValue(this.sumLong);
                 }
-                else if (this.valueType == typeof(double))
+                else
                 {
-                    return new DataValue(this.sumDouble);
+                    return new DataValue(this.sumDouble + this.sumLong);
                 }
-
-                throw new Exception("Unsupported Type");
             }
         }
 
-        public void Update<T>(T value)
+        public void Update<T>(Instrument instrument, T value)
             where T : struct
         {
+            bool isDeltaTemporality = true;
+
+            Type genType = instrument.GetType().GetGenericTypeDefinition();
+            if (genType == typeof(ObservableCounter<>) ||
+                genType == typeof(ObservableGauge<>))
+            {
+                isDeltaTemporality = false;
+            }
+
+            if (typeof(T) == typeof(long))
+            {
+                var val = (long)(object)value;
+
+                if (isDeltaTemporality)
+                {
+                    this.UpdateDeltaLong(val);
+                }
+                else
+                {
+                    this.UpdateCumulativeLong(val);
+                }
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                var val = (double)(object)value;
+
+                if (isDeltaTemporality)
+                {
+                    this.UpdateDeltaDouble(val);
+                }
+                else
+                {
+                    this.UpdateCumulativeDouble(val);
+                }
+            }
+            else
+            {
+                throw new Exception("Unsupported Type!");
+            }
+        }
+
+        public IMetric Collect(DateTimeOffset dt, bool isDelta)
```

Comment:
after reading the aggregator spec, and this PR - I don't think we should accept isDelta parameter in the Collect() call. An aggregator can produce either delta or cumulative, and that is determined in the ctor() itself, and cannot be changed afterwards. If we mix and match Collect() calls with isDelta=true,false, then this won't give the correct results.

As captured in this PR, the aggregators can be grouped into 2 categories
1. DeltaAggregators which output delta metric.
2. CumulativeAggregators which output cumulative metric.

DeltaAggregators can accept delta/cumulative inputs.
1.a DeltaAggregator accepting delta input
1.b DeltaAggregator accepting cumulative input

2.a CumulativeAggregators accepting delta input
2.b CumulativeAggregators accepting cumulative input

All the above 4 cases are handled in this PR, but I'd suggest to make aggregator component do less and have SDK/AggregatorStore part do more. Right now, aggregators accept the Instrument itself as input and makes decision based on instrument types. We should move this to AggStore, and have it pair the right instrument to right aggregator, so that aggregator just need to do the simple math part, and nothing more.

1a.
Update(delta) - adds delta to the sum it maintains
Collect() - returns the current sum, and resets sum/start time.

1b. 
Update(cumulative) - stores cumulative as the sum.
Collect() - must do cumulative -> delta conversion. (maintain last sum, do the difference as done already in this PR)

2a. 
Update(delta) - adds delta to the sum it maintains
Collect() - returns the current sum, (no reset)

2b. 

Update(cumulative) - stores cumulative as the sum
Collect() -return the current sum (no reset)